SYCL/Basic/half_builtins.cpp

```diff
@@ -0,0 +1,227 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+
+// OpenCL CPU driver does not support cl_khr_fp16 extension
+// UNSUPPORTED: cpu && opencl
+
+#include <CL/sycl.hpp>
+
+#include <cmath>
+#include <unordered_set>
+
+using namespace cl::sycl;
+
+constexpr int N = 16 * 3; // divisible by all vector sizes
+
+bool check(half a, half b) {
+  return fabs(2 * (a - b) / (a + b)) <
+             std::numeric_limits<cl::sycl::half>::epsilon() ||
+         a < std::numeric_limits<cl::sycl::half>::min();
+}
+
+#define TEST_BUILTIN_1_VEC_IMPL(NAME, SZ)                                      \
+  {                                                                            \
+    buffer<half##SZ> a_buf((half##SZ *)&a[0], N / SZ);                         \
+    buffer<half##SZ> d_buf((half##SZ *)&d[0], N / SZ);                         \
+    q.submit([&](handler &cgh) {                                               \
+      auto A = a_buf.get_access<access::mode::read>(cgh);                      \
+      auto D = d_buf.get_access<access::mode::write>(cgh);                     \
+      cgh.parallel_for(N / SZ,                                                 \
+                       [=](id<1> index) { D[index] = NAME(A[index]); });       \

```

Comment:
I think it would be more efficient to have a kernel testing all functions instead of a kernel per function.

I don't know how to apply the same logic for different vector length (i.e. reduce # of test kernels), but we probably can consider allocating more buffers (e.g. one buffer per vector length).